### DISCUSSION TITLE
v2.0.0 - Data-centric AI Ready

### DISCUSSION THREAD

**Discussion by cgnorthcutt** [cleanlab/cleanlab#238] [General]
# If you liked cleanlab v1.0.1, v2.0.0 will blow your mind! 💥🧠

cleanlab 2.0 adds powerful new workflows and algorithms for data-centric AI, dataset curation, auto-fixing label issues in data, learning with noisy labels, and more. Nearly every module, method, parameter, and docstring has been touched by this release.

If you're coming from 1.0, here's a [migration guide](https://docs.cleanlab.ai/master/migrating/migrate_v2.html).

### A few highlights of new functionalities in cleanlab 2.0:

1. rank every data point by label quality
2. find label issues in any dataset.
3. train any classifier on any dataset with label issues.
4. find overlapping classes to merge and/or delete at the dataset-level
5. yield an overall dataset health

For an in-depth overview of what cleanlab 2.0 can do, check out [this tutorial](https://docs.cleanlab.ai/master/tutorials/indepth_overview.html).

### To help you get started with 2.0, we've added:

* [extensive documentation](https://docs.cleanlab.ai/)
* [tutorials](https://docs.cleanlab.ai/master/tutorials/image.html)
* [updated examples](https://github.com/cleanlab/examples)
* [powerful new workflows](https://docs.cleanlab.ai/master/migrating/migrate_v2.html)
* [blogs](https://cleanlab.ai/blog/)

# Change Log 

This list is non-exhaustive! Assume every aspect of API has changed.

## Module name changes or moves:
- `classification.LearningWithNoisyLabels` class --> `classification.CleanLearning` class
- `pruning.py` --> `filter.py`
- `latent_estimation.py` --> `count.py`
- `cifar_cnn.py` --> `experimental/cifar_cnn.py`
- `coteaching.py` --> `experimental/coteaching.py`
- `fasttext.py` --> `experimental/fasttext.py`
- `mnist_pytorch.py` --> `experimental/fmnist_pytorch.py`
- `noise_generation.py` --> `benchmarking/noise_generation.py`
- `util.py` --> `internal/util.py`
- `latent_algebra.py` --> `internal/latent_algebra.py`

## Module Deletions:
- removed `polyplex.py`
- removed models/` --> (moved content to experimental/)

## New module created:
- `rank.py`
  - moved all ranking and ordering functions from `pruning.py`/`filter.py` to here
- `dataset.py`
  - brand new module supporting methods for dealing with data-level issues
- `benchmarking.py`
  - Future benchmarking modules go here. Moved `noise_generation.py` here.

## Method name changes:
- `pruning.get_noise_indices()` --> `filter.find_label_issues()`
- `count.num_label_errors()` --> `count.num_label_issues()`

## Methods added:
- `rank.py` adds
  - two ranking functions to rank data based on label quality for entire dataset (not just examples with label issues)
  - `get_self_confidence_for_each_label()`
  - `get_normalized_margin_for_each_label()`
- `filter.py` adds
  - two more methods added to `filter.find_label_issues()` (select method using the `filter_by` parameter)
    - `confident_learning`, which has been shown to work very well and may become the default in the future, and
    - `predicted_neq_given`, which is useful for benchmarking a simple baseline approach, but underperformant relative to the other filter_by methods)
- `classification.py` adds
  - `ClearnLearning.get_label_issues()`
    - for a canonical one-line of code use:`CleanLearning().fit(X, y).get_label_issues()` 
    - no need to compute predicted probabilities in advance
  - `CleanLearning.find_label_issues()`
    - returns a dataframe with label issues (instead of just a mask)

## Naming conventions changed in method names, comments, parameters, etc.
- `s` -> `labels` 
- `psx` -> `pred_probs`
- `label_errors` --> `label_issues`
- `noise_mask` --> `label_issues_mask`
- `label_errors_bool` --> `label_issues_mask`
- `prune_method` --> `filter_by`
- `prob_given_label` --> `self_confidence`
- `pruning` --> `filtering`

## Parameter re-ordering:
- re-ordered (`labels`, `pred_probs`) parameters to be consistent (in that order) in all methods.
- re-ordered parameters (e.g. `frac_noise`) in filter.find_label_issues()

## Parameter changes:
- in `order_label_issues()`
  - param: `sorted_index_method` --> `rank_by`
- in `find_label_issues()`
  - param: `sorted_index_method` --> `return_indices_ranked_by`
  - param: `prune_method` --> `filter_by`

## Global variables changed:
- `filter.py`
  - Only require 1 example to be left in each class
  - `MIN_NUM_PER_CLASS = 5` --> `MIN_NUM_PER_CLASS = 1`
  - enables cleanlab to work for toy-sized datasets

## Dependencies added
- pandas=1.0.0

## Way-too-detailed Change Log
* convert readme to markdown for pypi release. by @cgnorthcutt in https://github.com/cleanlab/cleanlab/pull/126
* Add EditorConfig by @anishathalye in https://github.com/cleanlab/cleanlab/pull/129
* Major API change. Introducing Cleanlab 2.0 by @cgnorthcutt in https://github.com/cleanlab/cleanlab/pull/128
* Standardize code style to Black by @anishathalye in https://github.com/cleanlab/cleanlab/pull/107
* Redirect RTD site to docs.cleanlab.ai by @weijinglok in https://github.com/cleanlab/cleanlab/pull/130
* Redirect RTD site to docs.cleanlab.ai: Part 2 by @weijinglok in https://github.com/cleanlab/cleanlab/pull/132
* Add image classification tutorial and streamline docs CI/CD by @weijinglok in https://github.com/cleanlab/cleanlab/pull/127
* remove redundant text by @jwmueller in https://github.com/cleanlab/cleanlab/pull/134
* Remove extra slashes in docs relative path by @weijinglok in https://github.com/cleanlab/cleanlab/pull/135
* Fix docs TOC for v2.0 by @weijinglok in https://github.com/cleanlab/cleanlab/pull/136
* Add label quality scoring functions and user API to choose the method by @JohnsonKuan in https://github.com/cleanlab/cleanlab/pull/131
* Change cleanlab version in Image Tutorial by @weijinglok in https://github.com/cleanlab/cleanlab/pull/138
* Utilites -> internal submodule refactor by @jwmueller in https://github.com/cleanlab/cleanlab/pull/141
* Fix NumPy deprecation warning by @anishathalye in https://github.com/cleanlab/cleanlab/pull/142
* Remove unnecessary print statement by @anishathalye in https://github.com/cleanlab/cleanlab/pull/145
* Add explanation that estimators must be clonable by @anishathalye in https://github.com/cleanlab/cleanlab/pull/146
* Update doc site quickstart page to reflect v2.0 API by @weijinglok in https://github.com/cleanlab/cleanlab/pull/143
* Fix sklearn estimator cloning by @anishathalye in https://github.com/cleanlab/cleanlab/pull/144
* Update default label quality scoring method to self_confidence by @JohnsonKuan in https://github.com/cleanlab/cleanlab/pull/147
* Allow n-dim data in LearningWithNoisyLabels by @anishathalye in https://github.com/cleanlab/cleanlab/pull/148
* Improve user-control by @jwmueller in https://github.com/cleanlab/cleanlab/pull/149
* Enable use of find_label_issues_kwargs for hyper-parameter search by @JohnsonKuan in https://github.com/cleanlab/cleanlab/pull/152
* Add fix and test for sklearn GridSearchCV with LearningWithNoisyLabels by @JohnsonKuan in https://github.com/cleanlab/cleanlab/pull/153
* Add tutorial for tabular data classification by @weijinglok in https://github.com/cleanlab/cleanlab/pull/151
* Minor tutorial edits by @jwmueller in https://github.com/cleanlab/cleanlab/pull/155
* Add Python 3.10 to CI by @anishathalye in https://github.com/cleanlab/cleanlab/pull/160
* Add development guide by @anishathalye in https://github.com/cleanlab/cleanlab/pull/164
* Add text classification tutorial by @weijinglok in https://github.com/cleanlab/cleanlab/pull/154
* Add audio tutorial to doc site by @weijinglok in https://github.com/cleanlab/cleanlab/pull/165
* Add overview for computing out-of-sample predicted probabilities with cross-validation to doc site by @weijinglok in https://github.com/cleanlab/cleanlab/pull/166
* Add CI check that .ipynb outputs are empty by @anishathalye in https://github.com/cleanlab/cleanlab/pull/169
* Add CI check for trailing newlines in notebooks by @anishathalye in https://github.com/cleanlab/cleanlab/pull/170
* Improve image tutorial accuracy and finding better label errors by @weijinglok in https://github.com/cleanlab/cleanlab/pull/167
* Remove unnecessary version warning by @anishathalye in https://github.com/cleanlab/cleanlab/pull/162
* Add test to check examples are found by cleanlab by @weijinglok in https://github.com/cleanlab/cleanlab/pull/172
* Various tutorial improvements by @jwmueller in https://github.com/cleanlab/cleanlab/pull/173
* Deploys docs only if triggered by master branch by @weijinglok in https://github.com/cleanlab/cleanlab/pull/175
* added LearningWithNoisyLabels.find_label_issues instance method by @jwmueller in https://github.com/cleanlab/cleanlab/pull/157
* Add note on EditorConfig to development guide by @anishathalye in https://github.com/cleanlab/cleanlab/pull/176
* CleanLearning = Machine Learning with cleaned data by @cgnorthcutt in https://github.com/cleanlab/cleanlab/pull/177
* Simple fix to Issue 158 (and potentially other issues) by @cgnorthcutt in https://github.com/cleanlab/cleanlab/pull/178
* Update docs README by @weijinglok in https://github.com/cleanlab/cleanlab/pull/180
* Polish the APIs and file-structure to prepare for 2.0 release by @jwmueller in https://github.com/cleanlab/cleanlab/pull/181
* Make EditorConfig match Jupyter for notebooks by @anishathalye in https://github.com/cleanlab/cleanlab/pull/179
* Add figure to out-of-sample pred proba via cv tutorial by @weijinglok in https://github.com/cleanlab/cleanlab/pull/183
* Move rest of example_models/ -> experimental/ by @jwmueller in https://github.com/cleanlab/cleanlab/pull/184
* Fix typo in test by @anishathalye in https://github.com/cleanlab/cleanlab/pull/186
* Add more ergonomic method to skip notebooks by @anishathalye in https://github.com/cleanlab/cleanlab/pull/185
* Add link checking for all Markdown files by @anishathalye in https://github.com/cleanlab/cleanlab/pull/187
* Add link checking for compiled docs by @anishathalye in https://github.com/cleanlab/cleanlab/pull/188
* Minor improvements to count docstring by @cgnorthcutt in https://github.com/cleanlab/cleanlab/pull/190
* Add a GitHub icon and link below the docs' project title by @weijinglok in https://github.com/cleanlab/cleanlab/pull/192
* Add self.labels as attribute of FastTextClassifier by @JohnsonKuan in https://github.com/cleanlab/cleanlab/pull/194
* Move noise_generation into benchmarking module by @anishathalye in https://github.com/cleanlab/cleanlab/pull/196
* Remove import of internal package by @anishathalye in https://github.com/cleanlab/cleanlab/pull/195
* more specific filter-warning by @jwmueller in https://github.com/cleanlab/cleanlab/pull/193
* Remove deprecated functions by @anishathalye in https://github.com/cleanlab/cleanlab/pull/197
* Docs cleanup by @anishathalye in https://github.com/cleanlab/cleanlab/pull/189
* Introducing the new Dataset Module for cleanlab 2.0 by @cgnorthcutt in https://github.com/cleanlab/cleanlab/pull/182
* Readme reformat by @jwmueller in https://github.com/cleanlab/cleanlab/pull/198
* Returns DataFrame type from CleanLearning functions by @jwmueller in https://github.com/cleanlab/cleanlab/pull/199
* Migration guide for v2 by @jwmueller in https://github.com/cleanlab/cleanlab/pull/200
* set verbose default false. fix order of printing by @cgnorthcutt in https://github.com/cleanlab/cleanlab/pull/204
* Add from . import dataset to init by @cgnorthcutt in https://github.com/cleanlab/cleanlab/pull/205
* Make minor doc tweaks by @anishathalye in https://github.com/cleanlab/cleanlab/pull/203
* Fix error in docstring. missing item in tuple by @cgnorthcutt in https://github.com/cleanlab/cleanlab/pull/206
* fix broken printing of matrices by @cgnorthcutt in https://github.com/cleanlab/cleanlab/pull/207
* Add in-depth tutorial [WIP]  by @weijinglok in https://github.com/cleanlab/cleanlab/pull/208
* Revise migration guide by @anishathalye in https://github.com/cleanlab/cleanlab/pull/209
* Tutorial header-levels decreased by @jwmueller in https://github.com/cleanlab/cleanlab/pull/210
* Dark plots by @jwmueller in https://github.com/cleanlab/cleanlab/pull/211
* bug fixes and jupyter notebook support added by @cgnorthcutt in https://github.com/cleanlab/cleanlab/pull/212
* Add spaces at the end of doc side nav toc by @weijinglok in https://github.com/cleanlab/cleanlab/pull/213
* Clarify and fix several docstrings. by @cgnorthcutt in https://github.com/cleanlab/cleanlab/pull/214
* Configure display setting of Dataframe in 2.0 tutorial by @JohnsonKuan in https://github.com/cleanlab/cleanlab/pull/215
* Formats dataset health tutorial by @jwmueller in https://github.com/cleanlab/cleanlab/pull/216
* final tutorial edits. dataset docstrings imp by @cgnorthcutt in https://github.com/cleanlab/cleanlab/pull/217
* Deploy docs when new release is tagged by @weijinglok in https://github.com/cleanlab/cleanlab/pull/219
* Hardcode v1.0.1 hyperlink in in doc site by @weijinglok in https://github.com/cleanlab/cleanlab/pull/221
* Make dataset tutorial runnable on website docs, improve pulldown formatting by @jwmueller in https://github.com/cleanlab/cleanlab/pull/220
* Final docs polishing patches by @jwmueller in https://github.com/cleanlab/cleanlab/pull/223
* Bump version to 2.0.0 by @jwmueller in https://github.com/cleanlab/cleanlab/pull/222
* fix black formatting compliance by @cgnorthcutt in https://github.com/cleanlab/cleanlab/pull/224
* Update readme links for v2.0 by @jwmueller in https://github.com/cleanlab/cleanlab/pull/225
* Remove unneeded alt text by @anishathalye in https://github.com/cleanlab/cleanlab/pull/228
* Add tutorial toc page by @weijinglok in https://github.com/cleanlab/cleanlab/pull/230
* Proofreading README.md by @calebchiam in https://github.com/cleanlab/cleanlab/pull/226
* Generalize text tutorial to multiclass datasets by @calebchiam in https://github.com/cleanlab/cleanlab/pull/229
* Make small fixes by @anishathalye in https://github.com/cleanlab/cleanlab/pull/235
* Address tutorial points of confusion by @jwmueller in https://github.com/cleanlab/cleanlab/pull/233

## New Contributors
* @JohnsonKuan made their first contribution in https://github.com/cleanlab/cleanlab/pull/131
* @calebchiam made their first contribution in https://github.com/cleanlab/cleanlab/pull/226

**Full Changelog**: https://github.com/cleanlab/cleanlab/compare/v1.0.1...v2.0.0

<hr /><em>This discussion was created from the release <a href='https://github.com/cleanlab/cleanlab/releases/tag/v2.0.0'>v2.0.0 - Data-centric AI Ready</a>.</em>
